### PR TITLE
feat(SDRS): add a new resource sdrs protected instance

### PR DESCRIPTION
**Pull request by deer-hang**
**What this PR does / why we need it**:
add a new resource sdrs protected instance
**Which issue this PR fixes**:
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
fixes #xxx

**Special notes for your reviewer**:
- This PR depends on : feat(SDRS): add a new resource sdrs protection group #3117.
- Make sure the PR #3117 has been merged before merge this PR.

**Release note**:

```release-note

```

## PR Checklist

* [X] Tests added/passed.
* [X] Documentation updated.
* [X] Schema updated.

## Acceptance Steps Performed

```
make testacc TEST='./huaweicloud/services/acceptance/sdrs' TESTARGS='-run TestAccProtectedInstance_basic'

==> Checking that code complies with gofmt requirements... 
TF_ACC=1 go test ./huaweicloud/services/acceptance/sdrs -v -run TestAccProtectedInstance_basic -timeout 360m -parallel 4 
=== RUN   TestAccProtectedInstance_basic 
=== PAUSE TestAccProtectedInstance_basic
=== CONT  TestAccProtectedInstance_basic
--- PASS: TestAccProtectedInstance_basic (388.59s) 
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/sdrs      388.648s
```

### CODE REVIEW

File: docs/resources/sdrs_protectedinstance.md [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+  image_id           = data.huaweicloud_images_image.test.id
+  flavor_id          = data.huaweicloud_compute_flavors.test.ids[0]
+  security_group_ids = [data.huaweicloud_networking_secgroup.test.id]
+  availability_zone  = data.huaweicloud_availability_zones.test.names[0]
+  system_disk_type   = "SSD"
+
+  network {
+    uuid = data.huaweicloud_vpc_subnet.test.id
+  }
+}
```
Comment (Lance52259) [inline docs/resources/sdrs_protectedinstance.md]: pls use variables

File: docs/resources/sdrs_protectedinstance.md [hunk 2]
```diff
@@ -0,0 +1,181 @@
+---
+subcategory: "Storage Disaster Recovery Service (SDRS)"
+---
+
+# huaweicloud_sdrs_protectedinstance
```
Comment (Lance52259) [inline docs/resources/sdrs_protectedinstance.md]:
```suggestion
# huaweicloud_sdrs_protected_instance
```

File: huaweicloud/services/sdrs/resource_huaweicloud_sdrs_protectedinstance.go [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+			"tags": common.TagsSchema(),
+			"target_server": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceSDRSProtectedInstanceCreate(ctx context.Context, d *schema.ResourceData,
```
Comment (Lance52259) [inline huaweicloud/services/sdrs/resource_huaweicloud_sdrs_protectedinstance.go]: SDRS is unnecessary because of  the package is already names `sdrs`

**Comment by deer-hang** [comment]
```
make testacc TEST='./huaweicloud/services/acceptance/sdrs' TESTARGS='-run TestAccProtectedInstance_basic'
==> Checking that code complies with gofmt requirements... 
TF_ACC=1 go test ./huaweicloud/services/acceptance/sdrs -v -run TestAccProtectedInstance_basic -timeout 360m -parallel 4 
=== RUN   TestAccProtectedInstance_basic 
=== PAUSE TestAccProtectedInstance_basic
=== CONT  TestAccProtectedInstance_basic
--- PASS: TestAccProtectedInstance_basic (388.59s) 
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/sdrs      388.648s
```

File: huaweicloud/services/sdrs/resource_huaweicloud_sdrs_protected_instance.go [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+		return diag.Errorf("error creating SDRS client: %s", err)
+	}
+
+	n, err := protectedinstances.Get(client, d.Id()).Extract()
+	if err != nil {
+		if errCode, ok := err.(golangsdk.ErrDefault400); ok {
+			if resp, pErr := common.ParseErrorMsg(errCode.Body); pErr == nil && resp.ErrorCode == "SDRS.0208" {
+				// `SDRS.0208` means invalid protected instance ID
+				return common.CheckDeletedDiag(d, golangsdk.ErrDefault404{},
+					"error retrieving SDRS protected instance")
```
Comment (chengxiangdong) [inline huaweicloud/services/sdrs/resource_huaweicloud_sdrs_protected_instance.go:151]:
```suggestion
				return diag.Errorf("invalid protected instance ID: %s", d.Id())
```

File: huaweicloud/services/sdrs/resource_huaweicloud_sdrs_protected_instance.go [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+			}
+		}
+		return diag.FromErr(err)
```
Comment (chengxiangdong) [inline huaweicloud/services/sdrs/resource_huaweicloud_sdrs_protected_instance.go:154]:
```suggestion
		return common.CheckDeletedDiag(d, err, "error retrieving SDRS protected instance")
```

File: huaweicloud/services/sdrs/resource_huaweicloud_sdrs_protected_instance.go [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+	n, err := protectedinstances.Delete(client, d.Id(), deleteOpts).ExtractJobResponse()
+	if err != nil {
+		if errCode, ok := err.(golangsdk.ErrDefault400); ok {
+			if resp, pErr := common.ParseErrorMsg(errCode.Body); pErr == nil && resp.ErrorCode == "SDRS.0208" {
+				// `SDRS.0208` means invalid protected instance ID
+				return common.CheckDeletedDiag(d, golangsdk.ErrDefault404{},
+					"error deleting SDRS protected instance")
+			}
+		}
+		return diag.FromErr(err)
```
Comment (chengxiangdong) [inline huaweicloud/services/sdrs/resource_huaweicloud_sdrs_protected_instance.go:226]:
```suggestion
		return common.CheckDeletedDiag(d, err, "error deletiing SDRS protected instance")
```

File: huaweicloud/services/sdrs/resource_huaweicloud_sdrs_protected_instance.go [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+		DeleteTargetServer: d.Get("delete_target_server").(bool),
+		DeleteTargetEip:    d.Get("delete_target_eip").(bool),
+	}
+	n, err := protectedinstances.Delete(client, d.Id(), deleteOpts).ExtractJobResponse()
+	if err != nil {
+		if errCode, ok := err.(golangsdk.ErrDefault400); ok {
+			if resp, pErr := common.ParseErrorMsg(errCode.Body); pErr == nil && resp.ErrorCode == "SDRS.0208" {
+				// `SDRS.0208` means invalid protected instance ID
+				return common.CheckDeletedDiag(d, golangsdk.ErrDefault404{},
+					"error deleting SDRS protected instance")
```
Comment (chengxiangdong) [inline huaweicloud/services/sdrs/resource_huaweicloud_sdrs_protected_instance.go:223]:
```suggestion
				return diag.Errorf("invalid protected instance ID: %s", d.Id())
```

Looks like it doesn't return `SDRS.0208` on delete

Comment (chengxiangdong) [reply huaweicloud/services/sdrs/resource_huaweicloud_sdrs_protected_instance.go:223]: If `SDRS.0208` is returned when deleting, what should be done to delete it?

File: huaweicloud/services/sdrs/resource_huaweicloud_sdrs_protected_instance.go [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+			}
+		}
+		return diag.FromErr(err)
+	}
+
+	deleteTimeoutSec := int(d.Timeout(schema.TimeoutDelete).Seconds())
+	if err := protectedinstances.WaitForJobSuccess(client, deleteTimeoutSec, n.JobID); err != nil {
+		return diag.FromErr(err)
+	}
+	return nil
```
Comment (chengxiangdong) [inline huaweicloud/services/sdrs/resource_huaweicloud_sdrs_protected_instance.go:233]:
```suggestion
	d.SetId("")
	return nil
```

do not  forget to clean up the resources after deletion.

**Review by chengxiangdong — COMMENTED** [review]
The line width is 150, and some functions do not need to wrap.

File: huaweicloud/services/sdrs/resource_huaweicloud_sdrs_protected_instance.go (hunk 8 above)
Comment (deer-hang) [reply huaweicloud/services/sdrs/resource_huaweicloud_sdrs_protected_instance.go:233]: The deletion method will automatically set the ID to empty internally.

File: docs/resources/sdrs_protected_instance.md [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+## Argument Reference
+
+The following arguments are supported:
+
+* `region` - (Optional, String, ForceNew) Specifies the region in which to create the resource.
+  If omitted, the provider-level region will be used. Changing this parameter will create a new resource.
+
+* `name` - (Required, String) Specifies the name of a protected instance. The name can contain a maximum of 64
+  characters. The value can contain only letters (a to z and A to Z), digits (0 to 9), dots (.),
+  underscores (_), and hyphens (-).
```
Comment (Lance52259) [inline docs/resources/sdrs_protected_instance.md:39]:
```suggestion
  characters. The value can contain only letters, digits, dots (.), underscores (_) and hyphens (-).
```

File: huaweicloud/services/acceptance/sdrs/resource_huaweicloud_sdrs_protected_instance_test.go [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+  name        = "CentOS 7.6 64bit"
+  most_recent = true
+}
+
+data "huaweicloud_compute_flavors" "test" {
+  availability_zone = data.huaweicloud_availability_zones.test.names[0]
+  performance_type  = "normal"
+  cpu_core_count    = 2
+  memory_size       = 4
+}
```
Comment (Lance52259) [inline huaweicloud/services/acceptance/sdrs/resource_huaweicloud_sdrs_protected_instance_test.go:104]: use `common.TestBaseComputeResources` instead

Comment (deer-hang) [reply huaweicloud/services/acceptance/sdrs/resource_huaweicloud_sdrs_protected_instance_test.go:104]: The default image of `common.TestBaseComputeResources` is `Ubuntu 18.04 server 64bit`.
The image is not suit for this resource.

File: docs/resources/sdrs_protected_instance.md [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+```
+
+Note that the imported state may not be identical to your resource definition, due to some attributes missing from the
+API response, security or some other reason. The missing attributes include:
+`cluster_id`, `primary_subnet_id`, `primary_ip_address`, `delete_target_server` and `delete_target_eip`.
+It is generally recommended running `terraform plan` after importing a resource.
+You can then decide if changes should be applied to the resource, or the resource definition should be updated to align
+with the resource. Also, you can ignore changes as below.
+
+```bash
```
Comment (hmbig2) [inline docs/resources/sdrs_protected_instance.md]:
```suggestion
```hcl
```

**Comment by deer-hang** [comment]
```
make testacc TEST='./huaweicloud/services/acceptance/sdrs' TESTARGS='-run TestAccProtectedInstance_basic'
==> Checking that code complies with gofmt requirements... 
TF_ACC=1 go test ./huaweicloud/services/acceptance/sdrs -v -run TestAccProtectedInstance_basic -timeout 360m -parallel 4 
=== RUN   TestAccProtectedInstance_basic 
=== PAUSE TestAccProtectedInstance_basic
=== CONT  TestAccProtectedInstance_basic
--- PASS: TestAccProtectedInstance_basic (356.56s) 
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/sdrs      356.614s
```

File: docs/resources/sdrs_protected_instance.md (hunk 11 above)
Comment (deer-hang) [reply docs/resources/sdrs_protected_instance.md]: done

**Pull request merged**